### PR TITLE
Fix IndexStyle and abstract depwarns on 0.6

### PR DESCRIPTION
**Pull request by timholy**
When someone tags this, it should probably be a minor version bump since it introduces a new deprecation.

### CODE REVIEW

File: src/base-traits.jl [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
+export IsLeafType, IsBits, IsImmutable, IsContiguous, IsIndexLinear,
        IsAnything, IsNothing, IsCallable
 
 "Trait which contains all types"
 @traitdef IsAnything{X}
 @traitimpl IsAnything{X} <- (x->true)(X)
 
 "Trait which contains no types"
-typealias IsNothing{X} Not{IsAnything{X}}
+@compat IsNothing{X} = Not{IsAnything{X}}
```
Comment (mauro3) [inline src/base-traits.jl]: Is there no `const` needed?

Comment (timholy) [inline src/base-traits.jl]: It doesn't appear to be necessary (see Julia's NEWS.md file), but it doesn't seem to hurt either. Most but not all of them in https://github.com/JuliaLang/julia/pull/20500/files seem to use `const`, so I can add it if you prefer.

Comment (mauro3) [inline src/base-traits.jl]: That NEWS.md entry is a bit confusing.  Anyway, it does seem to make a difference (unless something changed on 0.6 in the last 5 days):
```julia
julia> V = Int
Int64

julia> f() = convert(V,5.0)
f (generic function with 1 method)

julia> @code_warntype f()
Variables:
  #self#::#f

Body:
  begin
      return (Main.convert)(Main.V, 5.0)::Any
  end::Any

julia> const VV = Int
Int64

julia> ff() = convert(VV,5.0)
ff (generic function with 1 method)

julia> @code_warntype ff()
Variables:
  #self#::#ff

Body:
  begin
      return $(Expr(:invoke, MethodInstance for convert(::Type{Int64}, ::Float64), :(Main.convert), :(Main.VV), 5.0))
  end::Int64
```
So, probably better add the `const`.

Comment (timholy) [inline src/base-traits.jl]: Parametric types seem different:
```julia
julia> MyVector{T} = AbstractArray{T, 1}
AbstractArray{T,1} where T

julia> f(a) = convert(MyVector, a)
f (generic function with 1 method)

julia> @code_warntype f([1,2,3])
Variables:
  #self#::#f
  a::Array{Int64,1}

Body:
  begin 
      return a::Array{Int64,1}
  end::Array{Int64,1}
```
but nevertheless I'll add the `const`.

**Comment by mauro3** [comment]
Thanks Tim!

**Comment by timholy** [comment]
OK to tag?

If you haven't set up your repos for [attobot](https://github.com/attobot/attobot), it does make life easier.

**Comment by mauro3** [comment]
Here it is: https://github.com/JuliaLang/METADATA.jl/pull/8049

(Sometimes I need more prodding than the gentle hint you made in your first post ;-)

File: src/base-traits.jl (hunk 1 above)
Comment (tkelman) [inline src/base-traits.jl]: the const is implicit for parametric type aliases, but IMO a useful signal about what the line is doing

**Pull request merged**